calculation/seistech_calc/seistech_calc/disagg/DisaggData.py

```diff
@@ -58,30 +59,47 @@ def __init__(
         exceedance: Optional[float] = None,
         mean_values: Optional[pd.Series] = None,
     ):
-        self.fault_disagg = fault_disagg.sort_values("contribution", ascending=False)
-        self.ds_disagg = ds_disagg.sort_values("contribution", ascending=False)
+        self._fault_disagg = fault_disagg.sort_values("contribution", ascending=False)
+        self._ds_disagg = ds_disagg.sort_values("contribution", ascending=False)
 
         self.site_info = site_info
         self.im = im
         self.im_value = im_value
         self.exceedance = exceedance
         self.mean_values = mean_values
 
+
+    @property
+    def fault_disagg_id_ix(self):
+        return self._fault_disagg
+
+    @property
+    def ds_disagg_id_ix(self):
+        return self._ds_disagg
+
+    @property
+    def fault_disagg_id(self):
+        raise NotImplementedError()
+
+    @property
+    def ds_disagg_id(self):
+        raise NotImplementedError()
+
     @property
     def total_contributions(self) -> pd.Series:
-        return self.fault_disagg.contribution.append(
-            pd.Series({"distributed_seismicity": self.ds_disagg.contribution.sum()})
+        return self.fault_disagg_id.contribution.append(
+            pd.Series({"distributed_seismicity": self.ds_disagg_id.contribution.sum()})
```

Comment:
Please correct me if I understood something wrong, I've seen @property decorators a few times but wasn't sure what exactly does, especially when it does something more than like what these two do, obviously these two seems just getters. [ref](https://www.programiz.com/python-programming/property) and by reading this, it seems basically it's a getter/setter pythonic way.
```python
property
    def fault_disagg_id_ix(self):
        return self._fault_disagg

    @property
    def ds_disagg_id_ix(self):
        return self._ds_disagg
```

Anyway, if so, wouldn't this be an error?
`self.fault_disagg_id.contribution.append` because `self.fault_disagg_id` is basically the property decorator right above, `def fault_disagg_id(self)` which is currently raising not implemented error?

Was it meant to be something like this?
```python
return self.fault_disagg_id_ix.contribution.append ...... "distributed_seismicity": self.ds_disagg_id_ix.contribution.sum()})
```

Reply:
As is it looks like it will raise a non implemented error. I'm happy to be educated otherwise.

Reply:
Great, I knew that was the case!
Thanks for the clarification :+1:

Reply:
It doens't actually result in a `NotImplementedError` as a DisaggData class should never be instantiated directly. Instead only BranchDisaggData and EnsembleDisaggData are instantiated, which both implement `fault_disagg_id` and `ds_disagg_id`. 

I have added a comment to these to indicate that these have to implement by sub-classes

Reply:
Oh yep. That makes sense. Would it be possible to mark these as abstract methods / class? I think python has a module for that?

Reply:
Hmh pretty sure I looked at that once and decided for some reason against it. Not sure, not keen on doing this now (since we don't do it anywhere else). But I do agree that it should be more obvious, I'll rename it to `BaseDisaggData` which I think is what we do with other pure BaseClasses.